File: implementations/flutter/design_patterns_flutter/lib/patterns/structural/proxy/real_subject.dart
import 'dart:async';
import 'product.dart';
import 'subject.dart';

class HttpProductService implements ProductService {
  int _hits = 0;

  @override
  Future<Product> getProduct(String id) async {
    _hits++;
    await Future.delayed(const Duration(milliseconds: 600));
    return Product(id: id, name: 'Product $id', price: 10 + double.tryParse(id)! * 1.5);
  }

  int getNetworkHits() => _hits;
}

File: implementations/react-native/design-patterns-rn/src/patterns/structural/proxy/real-subject.ts
import type { ProductService } from './subject';
import type { Product } from './type/type';

export class HttpProductService implements ProductService {
  private hits = 0;

  async getProduct(id: string): Promise<Product> {
    this.hits++;
    await sleep(600);
    return {
      id,
      name: `Product ${id}`,
      price: 10 + Number(id) * 1.5,
    };
  }

  getNetworkHits() { return this.hits; }
}

const sleep = (ms: number) => new Promise(res => setTimeout(res, ms));
